modules/tabs/test/spec.coffee

```diff
@@ -1,2 +1,212 @@
+    verifyVisible headers, 1
+
+    contentRenderer = tabs.component()._.options.contentRenderer
+    contentRenderer.should.have.been.called.with tabsContent.node(), {
+      name: 'Kate',
+      town: 'North Mymms'
+    }
+
+    spy.should.be.called.with { id: 1, value: 1, cause }
+
+  testContext = (tabs) ->
+    should.not.exist hx.palette.borderContext tabs.select '.hx-tabs-content'
+    tabs.component().select 2
+    ctx = hx.palette.borderContext tabs.select '.hx-tabs-content'
+    ctx.should.equal 'negative'
+
+  it 'should select the first element by default', ->
+    { headerAndBody } = setupTabs()
+    verifyVisible headerAndBody, 0
+
+  it 'should correctly emit the change event if changed from the api', (done) ->
+    tabToSelect = 1
+    { tabs } = testChangeEventFired 'api', tabToSelect, done
+    tabs.select tabToSelect
+
+  it 'should correctly select the tab if changed from the api', ->
+    tabToSelect = 1
+    { tabs, headerAndBody } = setupTabs()
+    tabs.select tabToSelect
+    verifyVisible headerAndBody, tabToSelect
+
+  it 'should correctly emit the change event if changed from the dom', (done) ->
+    tabToSelect = 1
+    { rootSel } = testChangeEventFired 'user', tabToSelect, done
+    pretendClickTab rootSel, tabToSelect
+
+  it 'should correctly select the tab if changed from the dom', ->
+    tabToSelect = 1
+    { tabs, rootSel, headerAndBody } = setupTabs()
+    pretendClickTab rootSel, tabToSelect
+    verifyVisible headerAndBody, tabToSelect
+
+  it 'should set the hx-tabs class', ->
+    tabs = setupTabsUsingStructuredObject()
+    tabs.classed('hx-tabs').should.equal true
+
+  it 'should set the hx-tabs-content element, if it does not exist already', ->
+    tabs = setupTabsUsingStructuredObject()
+    tabs.select('.hx-tabs-content').size().should.equal 1
+
+  it 'should select and render the first element by default', ->
+    tabs = setupTabsUsingStructuredObject()
+    headers = getHeaders tabs
+    tabsContent = tabs.select '.hx-tabs-content'
+    verifyVisible headers, 0
+
+    contentRenderer = tabs.component()._.options.contentRenderer
+    contentRenderer.should.have.been.called.with tabsContent.node(), {
+      name: 'Bob',
+      town: 'Wheathampstead'
+    }
+  
+  it 'should render the correct text for titles', ->
+    tabs = setupTabsUsingStructuredObject()
+    allTitles = getHeaders tabs
+      .map ({ header }) -> header.text()
+    allTitles.should.eql ['Example1', 'Example2', 'Example3', 'Example4']
+
+  it 'should allow things to be set by the api', ->
+    testChange 'api', (tabs) -> tabs.component().select 1
+
+  it 'should allow things to be set by the dom', ->
+    testChange 'user', (tabs) -> pretendClickTab tabs, 1
+
+  it 'should set the pallette correctly', ->
+    tabs = setupTabsUsingStructuredObject()
+    headers = tabs.selectAll '.hx-tab'
+    actualContext = headers.map (header) -> hx.palette.context header
+    expectedContext = [undefined, undefined, 'negative', 'action']
```

Comment:
Could we create a shared variable for this and `ctx` used in the setup function?